### PR TITLE
Integrate EPS (Electrum Personal Server)

### PR DESCRIPTION
**Pull request by maltokyo**
After the ElectrumX integration in the last release, I thought that some people may prefer the Electrum Personal Server (EPS) approach, which can only serve specific pre-determined wallets (using XPUBS). EPS cannot serve as a server for ad-hoc wallets with connections for many people, that ElectrumX delivers. The upside pf EPS is that it is much lighter, and doesn't require txindex=1.

I will integrate it with this pull request, but it might take a while, as I have some busy weeks ahead of me.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
awesome! yes lot's of people asked for it. I would prefer if we can get "official" docker images from @chris-belcher though

**Comment by chris-belcher** [comment]
I've never used docker before, so you're better off using a docker image that someone else creates.

**Comment by maltokyo** [comment]
We had a discussion here on Reddit about this:

https://www.reddit.com/r/Electrum/comments/dqzsxl/electrumx_from_docker_integrate_with_btcpay/?utm_source=share&utm_medium=ios_app&utm_name=iossmf


@peleion is currently checking with
@chris-belcher about this, to see if he could be recognized as the
“official” docker. I’m working with him, on the assumption it will be ;)

**Comment by peleion** [comment]
I've talked with @chris-belcher - we have a working Docker that's being stress tested now:

https://github.com/peleion/eps-docker
https://hub.docker.com/repository/docker/peleion/eps

@maltokyo and I will get it integrated. Ideally an install script would set it up out of the box but initially it is likely that it will have to be done manually but this only requires editing a few lines in a small config file and restarting the container a few times. 

Only thing left to do is pulling in existing wallets that have transactions already - shouldn't take long. Will report back . . .

**Comment by maltokyo** [comment]
@NicolasDorier - @peleion and I are working on it. 

However we might be stuck. Let me confirm something with you (or anyone else who knows for sure). 

I remember that BTCPay standard setup has the bitcoin node internal wallet turned off/deactivated by default. 

It seems that EPS is dependent on bitcoind in the sense that it uses/modifies its internal wallet - EPS doesn't store any database itself as confirmed with @chris-belcher.   So, if  BTCpay will not use bitcoind's wallet than EPS is a nonstarter. Am I getting it wrong?

**Comment by maltokyo** [comment]
Ah, now I see that from 4 days ago, the wallet is now available!  So @peleion we are good to go!  Wallet is enabled. Check out the recent changes here, now you can see why I thought that the wallet was disabled:

https://github.com/btcpayserver/btcpayserver-docker/commit/733e2e25b02408d3f84f2ed589eac2c4d6c16839#diff-52c18259fab0d9b17c90aa41c92b5764

**Comment by NicolasDorier** [comment]
ok cool

File: docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    container_name: electrum_ps
+    ports:
+      - 50002:50002
+    volumes:
+      - "electrum_ps_datadir:/data"
+    environment:
+      - DAEMON_URL=electrumx:1VmSUVGBuLNWvZl0LExRDW0tvl6196-47RfXIzS384g=@bitcoind:43782
+volumes:
+  electrum_ps_datadir:
+exclusive:
```
Comment (NicolasDorier) [inline docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml]: can you change it to `incompatible:` I just added that now.

File: docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  electrum_ps:
+    image: docker.io/peleion/eps
+    restart: unless-stopped
+    container_name: electrum_ps
+    ports:
+      - 50002:50002
+    volumes:
+      - "electrum_ps_datadir:/data"
+    environment:
+      - DAEMON_URL=electrumx:1VmSUVGBuLNWvZl0LExRDW0tvl6196-47RfXIzS384g=@bitcoind:43782
```
Comment (NicolasDorier) [inline docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml]: please use a different user/password.

**Comment by NicolasDorier** [comment]
Missing the doc. For example, how do I set my xpub there?

**Comment by peleion** [comment]
1. The xpub(s) and bitcoind credentials are hardcoded into the EPS config.cfg file. If we want to pass this information to EPS on install a script will have to be written to modify this file first and could possibly be added to the Docker container via a startup script.  Shouldn't be too difficult but I'm wondering if there is a privacy issue having these floating around as env vars. Possibly we could reuse the credentials when the user initially enters the xpubs into BTCpay on initial install if an option like "do you want to set up an Electrum Personal Server" is selected prior. 

2. EPS uses bitcoind's integrated wallet to watch addresses - it doesn't store any local database like ElectrumX so this will only work on the most recent BTCpay install with the wallet enabled as of last week. Otherwise the user will have to enable it manually for installs previous to then. It is unclear to me whether this will work on a non-Electrum wallet xpub and I don't have one to test.

3. EPS only needs to store 2 things persistently between restarts: the config.cfg file and any custom (i.e. LetsEncrypt or other) SSL scripts. The default storage is in /usr/local/etc/electrum-personal-server but I can easily move it in the Docker. I typically use /srv for this but the fragment above shows /data. Is there a preference or default? Once a consensus is reached I'll update the container.

4. On first time startup EPS will need bitcoind to rescan the blockchain for any existing transactions in the wallet if it is not empty and this will obviously be the case with existing installs/wallets. There is a minor issue with the EPS rescan - it requires user input - but I'm working with @chris-belcher on a solution. Right now the script eps-rescan on my github must be run manually first but it is kinda hacky ATM. ToDo is to automatically start something like this on initial EPS install, rescan from origin, then restart EPS in server mode. Doesn't take too long - 1 hour 20 minutes for one 2,000 transaction Electrum wallet  on moderately slow hardware - and I think adding user input to ask for a wallet birthdate is asking for problems.

**Comment by NicolasDorier** [comment]
hey @peleion  thanks for maintaining this image.

For `1.` we do that everywhere for all images, we don't have any other way to pass information.
With EPS, we can't possibly track xpub dynamically, this mean that the user will have to setup environment variable and start the docker-compose. IMHO this is fine, people using EPS are not noobs.

For `2.`, it does not matter to us, as long as you create a proper volume in the docker fragment.

For `4.`, yeah, if it is too hard to use people just don't use. I am surprised rescan take long, especially since scantxoutset now exists.

**Comment by chris-belcher** [comment]
> It is unclear to me whether this will work on a non-Electrum wallet xpub and I don't have one to test.

The answer is no. Electrum uses the bip32 version bytes in its own standard, the version bytes are used to detect the address script type. So for example taking the xpub from Ledger's web wallet app won't work when importing into Electrum or EPS.

See also: https://github.com/spesmilo/electrum-docs/blob/master/xpub_version_bytes.rst

Regarding rescan taking long and `scantxoutset`, in order to avoid address reuse the wallet needs to know which addresses have been used before even if their balance is zero now. `scantxoutset` can't find this information.

File: docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml (hunk 2 above)
Comment (maltokyo) [reply docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml]: Done - please check, is it what you meant?

File: docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml (hunk 1 above)
Comment (maltokyo) [reply docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml]: @NicolasDorier - Done - please check, is it what you meant?

**Comment by maltokyo** [comment]
@peleion , I have made updates requested by @NicolasDorier , is there anything else we need to do before starting testing?  Could you please review the content of the file from your side?

**Comment by maltokyo** [comment]
Ah I just saw this @peleion 

> 3. EPS only needs to store 2 things persistently between restarts: the config.cfg file and any custom (i.e. LetsEncrypt or other) SSL scripts. The default storage is in /usr/local/etc/electrum-personal-server but I can easily move it in the Docker. I typically use /srv for this but the fragment above shows /data. Is there a preference or default? Once a consensus is reached I'll update the container.

The storage for all the containers is unified in BTCPay server.  Would it be possible to leave this setup to be defined in the users yml file as we have it in my version here?

**Comment by peleion** [comment]
> Would it be possible to leave this setup to be defined in the users yml file as we have it in my version here

I can change the docker container to use /data (it's set to /srv right now hence your error) or the easier option is to change the yml to:

`    volumes:`
`      - "electrum_ps_datadir:/srv"`

Easy fix either way. 

Is the opt-add-electrum-ps.yml complete with all the vars you want to pass? If so I'll make a config that reads them automagically so the user doesn't have to manually edit the config file. If you've already made one just send it to me and I'll incorporate it into the image.

Does btcpay docker use namespace separation (I do)? The current docker runs as root and I let the daemon worry about it. If not I'll need the GID/PID that the bitcoind daemon runs as so I can make EPS do also for consistency and security.

Ive tested the newest EPS version docker and should have time to make these changes this week and push it to github/dockerhub.

**Comment by peleion** [comment]
Latest version of EPS docker is available utilizing my changes to the docker-compose.yml and passes all my tests. Documentation [here](https://github.com/peleion/btcpayserver-docker).

**Comment by NicolasDorier** [comment]
@maltokyo still in draft ?

**Comment by maltokyo** [comment]
Sorry. Have not recently had time to work on it. Hopefully over Christmas and New year.

**Comment by maltokyo** [comment]
This should be ready to go, testing was good on my side.  Thank you all, esp @peleion for the help!

**Comment by maltokyo** [comment]
Actually, @peleion for me server starts but when I look at the logs (trying to check how synching is going) I get this error.  

```
# docker logs generated_electrum_ps_1
INFO:2020-02-09 15:01:05,280: Starting Electrum Personal Server
INFO:2020-02-09 15:01:05,280: Logging to /tmp/electrumpersonalserver.log
WARNING:2020-02-09 15:01:05,347: Unable to find .cookie file, try setting `datadir` config
```

Also, Nothing at all is added to the _data dir, it is writing to another completely randomly named volume.

**Comment by peleion** [comment]
Make sure the docker image is up to date. @chris-belcher made a change for us with the latest version and the config file was renamed - I made some subtle logic changes to reflect these. The config.ini will not take docker vars directly so I had to sed them in.
```
docker rmi peleion/eps-docker
docker volume rm <eps volume name(s)>
```
and restart from scratch. The new docker works perfectly for me and the btcpay config changes I pushed to you should reflect them. If not I'll spool up a btcpay machine and check it.

**Comment by maltokyo** [comment]
Some of your changes didn't work for me. Had to remove whitespace etc. I did try on brand new server and new install of BTCPay. Please could you give it a shot. I can't work it out on my side. Still get error above.

**Comment by peleion** [comment]
Docker compose fragment points to old docker image name - I pushed fix to maltokyo master to avoid opening a separate pull here for this minimal fix. Just make sure to delete the old volumes and images.

**Comment by maltokyo** [comment]
Ah, silly me.  Indeed, I overlooked that.  Will try again now, sorry to bother.

**Comment by maltokyo** [comment]
Removed all old images and containers but now getting the below.  Ill need to check the connection credentials I guess... does anything stick out as obvious to you @peleion ?

> INFO:2020-02-09 18:06:21,102: Starting Electrum Personal Server
> INFO:2020-02-09 18:06:21,103: Logging to /srv/eps.log
> Traceback (most recent call last):
>   File "/usr/local/lib/python3.8/site-packages/electrumpersonalserver/server/jsonrpc.py", line 46, in queryHTTP
>     self.conn.request("POST", self.url, body, headers)
>   File "/usr/local/lib/python3.8/http/client.py", line 1230, in request
>     self._send_request(method, url, body, headers, encode_chunked)
>   File "/usr/local/lib/python3.8/http/client.py", line 1276, in _send_request
>     self.endheaders(body, encode_chunked=encode_chunked)
>   File "/usr/local/lib/python3.8/http/client.py", line 1225, in endheaders
>     self._send_output(message_body, encode_chunked=encode_chunked)
>   File "/usr/local/lib/python3.8/http/client.py", line 1004, in _send_output
>     self.send(msg)
>   File "/usr/local/lib/python3.8/http/client.py", line 944, in send
>     self.connect()
>   File "/usr/local/lib/python3.8/http/client.py", line 915, in connect
>     self.sock = self._create_connection(
>   File "/usr/local/lib/python3.8/socket.py", line 807, in create_connection
>     raise err
>   File "/usr/local/lib/python3.8/socket.py", line 796, in create_connection
>     sock.connect(sa)
> ConnectionRefusedError: [Errno 111] Connection refused
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/usr/local/bin/electrum-personal-server", line 8, in <module>
>     sys.exit(main())
>   File "/usr/local/lib/python3.8/site-packages/electrumpersonalserver/server/common.py", line 815, in main
>     bestblockhash[0] = rpc.call("getbestblockhash", [])
>   File "/usr/local/lib/python3.8/site-packages/electrumpersonalserver/server/jsonrpc.py", line 85, in call
>     response = self.queryHTTP(request)
>   File "/usr/local/lib/python3.8/site-packages/electrumpersonalserver/server/jsonrpc.py", line 66, in queryHTTP
>     self.conn.connect()
>   File "/usr/local/lib/python3.8/http/client.py", line 915, in connect
>     self.sock = self._create_connection(
>   File "/usr/local/lib/python3.8/socket.py", line 807, in create_connection
>     raise err
>   File "/usr/local/lib/python3.8/socket.py", line 796, in create_connection
>     sock.connect(sa)
> ConnectionRefusedError: [Errno 111] Connection refused
> INFO:2020-02-09 18:06:25,180: Starting Electrum Personal Server
> INFO:2020-02-09 18:06:25,181: Logging to /srv/eps.log
> ERROR:2020-02-09 18:06:25,451: Error with bitcoin json-rpc: JsonRpcConnectionError('authentication for JSON-RPC failed')

**Comment by peleion** [comment]
Above my (python) pay grade, I'm afraid. Perhaps @chris-belcher or some other python guru can chime in . . . . 

My [image](https://github.com/peleion/eps-docker) is based on Debian Buster and Docker host runs Fedora without problems, not sure what btcpay server uses or whether this is important.

**Comment by maltokyo** [comment]
Did these credientials login to bitcoin rpc for you ok?

```
rpc_user = eps
rpc_password = cmhKyeLSco35KVhecK3blYcHFx73xrJ1fOQtjtmiKks
```

**Comment by peleion** [comment]
Where did your credentials come from? This is what I get for user "eps"
```
[root@0018664c0da9 /]# rpcauth.py eps
String to be appended to bitcoin.conf:
rpcauth=eps:15297f94397b5e6f09e462aeafc86a79$52ecad4be39b0031b2005b1d0a4b7e8514e546d351996bfd4922a13ceec9632d
Your password:
a-tN15C6JZfxGF46dv5yxulc1tEvCVje7jjysw5U84o=
```
I suspect EPS is just failing ungracefully with improper login.

**Comment by maltokyo** [comment]
I did generate them properly, but a couple of months ago, strange.  I generated a new set now, and tested and it all worked.  So I will just merge your PR.  Thanks!

@NicolasDorier this is good to go now, please do next steps.

File: docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    ports:
+      - "50002:50002"
+    volumes:
+      - "electrum_ps_datadir:/srv"
+    environment:
+      - HOST=bitcoind
+      - PORT=43782
+      - RPC_USER=eps
+      - RPC_PASSWORD=cmhKyeLSco35KVhecK3blYcHFx73xrJ1fOQtjtmiKks
+      - XPUB=zpub_BLAHBLAHBLAH_Add_Yours_Here
```
Comment (NicolasDorier) [inline docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml]: this should be configurable via environment variable.

**Comment by NicolasDorier** [comment]
`XPUB` need to be configurable. Also you need to document the fragment in the README.
To pass the environment variable from the host to docker-compose, you need to add a line in `build.sh`.

**Comment by maltokyo** [comment]
Hi @NicolasDorier  I think I have done it right, please check if that will work.
Docs I will update later on.

**Comment by maltokyo** [comment]
@NicolasDorier I removed this from the fragment:

```
incompatible:
  - pruning
```

Because actually EPS (unlike ElectrumX) is indeed compatible and works with pruned server.  This is why it might be so useful to users of BTCPay, as I guess most people are on a VPS etc, and use pruned servers.

Details here:
https://github.com/chris-belcher/electrum-personal-server#how-is-this-different-from-other-electrum-servers-

**Comment by maltokyo** [comment]
And, @NicolasDorier, I have updated README.  Is there anything else to do?  itsumo-arigato...!

File: README.md [hunk 4]
```diff
@@ -159,7 +159,8 @@ Available `BTCPAYGEN_ADDITIONAL_FRAGMENTS` currently are:
...
+* [opt-add-electrum-ps](docker-compose-generator/docker-fragments/opt-add-electrum-ps.yml), to integrate Electrum Personal Server (EPS) with BTCPay (EPS is a  single-user alternative to the ElectrumX Server option above).  EPS will also use the BTCPay server's full bitcoin node for complete privacy when using your own Electrum wallet, for your own personal use (i.e. other users cannot use your server to verify transactions).  Also, the bitcoin option `-txindex` is NOT mandatory for EPS, and it will run on a pruned node (unlike ElectrumX).  You will need to add your XPUB/YPUB/ZPUB as an environment variable before enabling EPS for the first time.  See [full documentation](https://github.com/btcpayserver/btcpayserver-doc/edit/master/ElectrumPersonalServer.md) for details.
```
Comment (NicolasDorier) [inline README.md]: the full `documentation link` it should be `https://github.com/btcpayserver/btcpayserver-doc/blob/master/ElectrumPersonalServer.md`

**Comment by NicolasDorier** [comment]
@maltokyo 

1. Rename the environment variable from `BTCPAY_EPS_XPUB` to `EPS_XPUB`
2. Document the variable just above in the README, where we document all the environment variable.
3. Mention the variable in the `opt-add-` so people know where to look for it.

File: README.md (hunk 4 above)
Comment (maltokyo) [reply README.md]: Done

**Comment by NicolasDorier** [comment]
PR seems good to go. Will try.

**Comment by maltokyo** [comment]
Done everything you commented on @NicolasDorier please give it a go

**Comment by NicolasDorier** [comment]
Things work, I will probably replace the docker image later with one properly supporting multi-arch and not starting before bitcoind is ready.

**Comment by maltokyo** [comment]
> Things work, I will probably replace the docker image later with one properly supporting multi-arch and not starting before bitcoind is ready.

Great, I would be interested to learn how to do this (not start until bitcoind is ready), and  @peleion is there another image supporting multi-arch?)

**Comment by NicolasDorier** [comment]
So to do it properly for a Python project, you can check how I build docker-compose cross platform by inspiring from https://github.com/btcpayserver/docker-compose-builder

Then, you do work on this repo https://github.com/btcpayserver/dockerfile-deps

**Comment by NicolasDorier** [comment]
Note that building docker-compose is more difficult than EPS, there is no "bootloader" or shit like that.

**Comment by peleion** [comment]
Assuming the current pull passes checks on amd64 it shouldn't be a problem to add multi-arch to the image using xbuild - I didn't know btcpay was a thing on arm. I've seen _lots_ of problems with bitcoind on a pi, maybe the 4 will be better. Someone else will have to test - I don't have any arm hardware right now.

I don't know of any signal that bitcoind sends to announce it is ready - if one exists someone let me know and I'll incorporate it. Otherwise a startup delay could be hardcoded or something similar but this starts to smell kinda hacky (and one more thing to break). Is this added complexity worth it? What is the downside to letting EPS simply idle waiting for bitcoind to become available?

**Comment by NicolasDorier** [comment]
@peleion in BTCPay we have a system to build whatever architecture we want via docker and qemu.
I think in either case I need to adapt to btcpay case. We already use a solution to wait until the node is synched by writing a file in a location when the node is ready. Docker images just sleep loop over the file until it exists.

**Pull request merged**